dapr_agents/agents/agent/agent.py

```diff
@@ -1,235 +1,279 @@
         """
-        for tool_call in tool_calls:
-            function_name = tool_call.function.name
-            tool_id = tool_call.id
-            function_args = (
-                tool_call.function.arguments_dict
-            )  # Use the property to get dict
-
-            if not function_name:
-                logger.error(f"Tool call missing function name: {tool_call}")
-                continue
+        # Limiting concurrency to avoid overwhelming downstream systems
+        max_concurrent = 10
+        semaphore = asyncio.Semaphore(max_concurrent)
+
+        async def run_and_record(tool_call: ToolCall) -> ToolMessage:
+            """
+            Executes a single tool call, respecting the concurrency limit.
+            Appends the result to the persistent audit log.
+            If the function name is missing, returns a ToolMessage with error status and raises AgentError.
+            """
+            async with semaphore:
+                function_name = tool_call.function.name
+                tool_id = tool_call.id
+                function_args = tool_call.function.arguments_dict
+
+                if not function_name:
+                    error_msg = f"Tool call missing function name: {tool_call}"
+                    logger.error(error_msg)
+                    # Return a ToolMessage with error status and raise AgentError
+                    tool_message = ToolMessage(
+                        tool_call_id=tool_id,
+                        name="<missing>",
+                        content=error_msg,
+                        role="tool",
+                    )
+                    self.tool_history.append(tool_message)
+                    raise AgentError(error_msg)
 
-            try:
-                logger.info(f"Executing {function_name} with arguments {function_args}")
-                result = await self.tool_executor.run_tool(
-                    function_name, **function_args
-                )
-                tool_message = ToolMessage(
-                    tool_call_id=tool_id, name=function_name, content=str(result)
-                )
-                self.text_formatter.print_message(tool_message)
-                self.tool_history.append(tool_message)
-            except Exception as e:
-                logger.error(f"Error executing tool {function_name}: {e}")
-                raise AgentError(f"Error executing tool '{function_name}': {e}") from e
+                try:
+                    logger.debug(
+                        f"Executing {function_name} with arguments {function_args}"
+                    )
+                    result = await self.run_tool(function_name, **function_args)
+                    tool_message = ToolMessage(
+                        tool_call_id=tool_id, name=function_name, content=str(result)
+                    )
+                    # Printing the tool message for visibility
+                    self.text_formatter.print_message(tool_message)
```

Comment:
can we make this optional based on like a verbose flag or something? That way the tool message, if huge, wont take up the whole terminal output

Reply:
absolutely